### PR TITLE
CanBeNonNullable: fix false positives for parameterized types

### PR DESCRIPTION
**Pull request by dzirbel**
CanBeNonNullable reports that properties of parameterized types can be marked as non-nullable even when this is impossible, just because the parameterized type is itself nullable (i.e. does not inherit from a non-nullable type). Instead, the rule should only report cases where a property is explicitly marked nullable but does not need to be.

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt [hunk 1]
```diff
@@ -370,6 +370,56 @@ class CanBeNonNullableSpec(val env: KotlinCoreEnvironment) {
...
+
+        @Test
+        fun `does not report on properties of a parameterized type which must be nullable`() {
+            val code = """
+                class P<T : Any>(private val foo: T) {
+                    val a: T
+                        get() = foo
+                    val b: T? = foo.takeIf { Random.nextBoolean() }
+                    val c: T?
+                        get() = foo.takeIf { Random.nextBoolean() }
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:405]: I miss a test where `T : Any` and this rule find some issues.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt [hunk 2]
```diff
@@ -370,6 +370,56 @@ class CanBeNonNullableSpec(val env: KotlinCoreEnvironment) {
...
+
+            assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
+        }
+
+        @Test
+        fun `does not report on properties of a parameterized type which resolve to Nothing`() {
+            val code = """
+                class P<T> {
+                    val foo: T
+                        get() = error("")
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:417]: 🤔 Should we have a test to check what happen if Foo is declared as `T?`?

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt (hunk 1 above)
Comment (dzirbel) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:416]: Good point; added

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt (hunk 2 above)
Comment (dzirbel) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:428]: Also a good idea; added

**Comment by dzirbel** [comment]
Hm, I forced pushed two additional tests and codecov is now complaining it's under the coverage target? 🤔 Did I do something wrong? I'm not as familiar with PRs in GitHub as other tools

**Comment by BraisGabin** [comment]
> Did I do something wrong?

Yet another caveat on our CI configuration... We usualy don't pay attention to that (unless the number is near to 0). Probably we should do something about it to avoid this type of confusions.

@schalkms What do you think we could do here?

**Comment by schalkms** [comment]
I guess you @BraisGabin and @dzirbel mean the two codecov CI runs.
Codecov provides detekt with code coverage reports. One report yields the code coverage for the whole project. The other report called patch just yields the code coverage for the additions and changed lines of code. The patch report fails when the code coverage is below the 80 % target. 80 % constitutes  a sensible target for well tested code. However, this report is not 100 % accurate for every PR. Usually, I pay attention to this code coverage report during code reviews.
In this case, it's fine to merge with 75 %, I guess. Nevertheless, if you @dzirbel have an idea on how to test a currently uncovered path, we'd ask you to add tests for it.

> Yet another caveat on our CI configuration... We usualy don't pay attention to that (unless the number is near to 0). Probably we should do something about it to avoid this type of confusions.

Maybe, we could spend one or two sentences in the `Contributing.md` file. I don't know whether this addition helps. What do you think?

> @schalkms What do you think we could do here?

**Comment by dzirbel** [comment]
> I guess you @BraisGabin and @dzirbel mean the two codecov CI runs. Codecov provides detekt with code coverage reports. One report yields the code coverage for the whole project. The other report called patch just yields the code coverage for the additions and changed lines of code. The patch report fails when the code coverage is below the 80 % target. 80 % constitutes a sensible target for well tested code. However, this report is not 100 % accurate for every PR. Usually, I pay attention to this code coverage report during code reviews. In this case, it's fine to merge with 75 %, I guess. Nevertheless, if you @dzirbel have an idea on how to test a currently uncovered path, we'd ask you to add tests for it.

Right, if the report can be inaccurate that all makes sense. In this case, coverage was above 80% initially (I believe, since all the checks had passed IIRC) and I [force pushed](https://github.com/detekt/detekt/compare/b852f6559e558fba88c0e51bb858c6c09c917567..0f1f1ac484d54d6051b2e06e11d0875ad8b55d22) two additional tests, after which coverage apparently dropped to 75%, so I wanted to double check that my PR method was correct (maybe I'd messed up rebases or something).

Thanks both of you for the quick and helpful reviews on my PRs this weekend :)

**Comment by schalkms** [comment]
You are welcome 🤗

**Pull request merged**